File: 프로그래머스/4/43236. 징검다리/징검다리.py
def solution(distance, rocks, n):
    rocks.sort()
    rocks.append(distance)

    def getRemovedCount(dist):
        start = 0
        removed = 0
        for e in rocks:
            gap = e-start
        
            if (gap<dist):
                removed +=1
                continue
            start=e
        return removed
    start = 0
    end = distance
    answer = 0
    while(start<=end):
        mid = (start+end)//2
        count = getRemovedCount(mid)
        if(count<=n):
            start = mid+1
            answer = max(answer, mid)
        else:
            end= mid-1
            

                
    return answer

File: 프로그래머스/4/43236. 징검다리/징검다리.js
function solution(last, rocks, N) {
    let answer = 0;
    rocks.sort((a,b)=>a-b)
    rocks = [...rocks,last]
    const M = rocks.length
    let [left, right] = [1, last]
    // console.log(rocks)
    function getRemoved(target){
        let [start, count] = [0,0]
        for(let i=0;i<M;i++){
            const gap = rocks[i]-start
            // console.log(start, rocks[i], count)
            if(gap>=target){start = rocks[i]}
            else{count+=1}
        }
        
        return count
    }
    while(left<=right){
        const mid = Math.ceil((left+right)/2)
        const count = getRemoved(mid)
        // console.log("####",count, mid)
        if(count>N){
            right = mid-1
        }else{
            left=mid+1
            answer = mid
            
        }
    }
    return answer;
}